homeassistant/components/recorder/statistics.py

```diff
@@ -1846,11 +1877,42 @@ def get_latest_short_term_statistics(
             )
         if not metadata:
             return {}
-        metadata_ids = _extract_metadata_and_discard_impossible_columns(metadata, types)
-        stmt = _latest_short_term_statistics_stmt(metadata_ids)
+        metadata_ids = set(
+            _extract_metadata_and_discard_impossible_columns(metadata, types)
+        )
+        stmt = _latest_short_term_statistics_from_latest_statistics_short_term_ids_stmt(
+            metadata_ids
+        )
         stats = cast(
             Sequence[Row], execute_stmt_lambda_element(session, stmt, orm_rows=False)
         )
+        found_metadata_ids = {stat.metadata_id for stat in stats}
+        # If we do not have a row for a metadata_id, we need to find it in the table
+        # manually with a slower query. This should only happen once per metadata_id
+        # ever since the latest statistics table was introduced.
+        if missing_metadata_ids := metadata_ids - found_metadata_ids:
+            missing_ids: set[int] = set()
+            for metadata_id in missing_metadata_ids:
+                if (
+                    latest := cast(
+                        Sequence[Row],
+                        execute_stmt_lambda_element(
+                            session,
+                            _find_latest_short_term_statistic_stmt(metadata_id),
+                            orm_rows=False,
+                        ),
+                    )
+                ) and (missing_id := latest[0].id):
+                    missing_ids.add(missing_id)
+            if missing_ids:
+                stmt = _short_term_statistics_by_ids_stmt(missing_ids)
+                if stats_for_missing_rows := cast(
+                    Sequence[Row],
+                    execute_stmt_lambda_element(session, stmt, orm_rows=False),
+                ):
+                    stats = list(stats)
+                    stats.extend(stats_for_missing_rows)
```

Comment:
Need to do some profiling to make sure we rarely hit this block in actual production

Reply:
Looks solid. py-spy below